### ISSUE TITLE
Administratively abandoning other player's claims adjusts the command executor's claim blocks

### ISSUE THREAD

**Issue by TomLewis** [GriefPrevention/GriefPrevention#322]
I just cant catch a break with GP randomly messing with data in the database after the fiasco with #279.
I haven't used Plan on my server since that event.
Well I'm back with another "what the hell is GP randomly doing", put on your "WTF hat", I cant figure it out. This time around 171105  claim blocks were added to an admins "accruedblocks" MySQL data after he noticed after removing a claim from a player.

The only thing I have changed recently is I have updated Paperspigot to the latest release:
PaperSpigot:
`This server is running Paper version git-Paper-1509 (MC: 1.12.2) (Implementing API version 1.12.2-R0.1-SNAPSHOT) Previous version: git-Paper-1476 (MC: 1.12.2)`
I updated Java & the OS and Waterfall bungee, I haven't touched the plugin in a long time.
GriefPrevention version 16.8.dev 

An admin deleted a claim of a user that's been offline for 745 days. The message back to him was him "_Claim abandoned.  You now have 171108 available claim blocks._" it wasn't his claim, yet it somehow looks like it changed his "accruedblocks" claim blocks on the surface, one thing I did find out is he used  /AbandonClaim instead of /deleteclaim so this may be where the issue arises, I'm not sure how to find out. He did not have 171105, he had 0, which is why he noticed.

We pulled his personal logs off his computer for today and found he had luckily run /claimslist earlier on in the day, and I run it again to compare the two to make sure none of his Claims had been resized or deleted, and you can clearly see the the difference between the two https://www.diffchecker.com/p8QmCFV4 the left is earlier on in the day, the right is after he noticed the massive increase in "accruedblocks".

After digging through his personal client logs, the server logs and GP's logs I found this in the GP logs:
```
13:35 Retrieved data for 7e556548-0a59-4c43-bd7b-44f3bc9b3261: accrued: 190751 bonus: 171133
13:35 Retrieved data for 7e556548-0a59-4c43-bd7b-44f3bc9b3261: accrued: 190751 bonus: 171133
13:35 Lawzoneon has more claimed land than blocks available.  Adding blocks to fix.
13:35 Lawzoneon Accrued blocks: 190685 Bonus blocks: 171133
13:35 Total blocks: 361818 Total claimed area: 361874
13:35 PirateCraft_1: x-3914, z1782 // PirateCraft_1: x-3819, z1879 = 9408
13:35 PirateCraft_1_nether: x-446, z228 // PirateCraft_1_nether: x-436, z243 = 176
13:35 PirateCraft_1: x2936, z-7639 // PirateCraft_1: x3092, z-7484 = 24492
13:35 PirateCraft_1: x-3683, z2242 // PirateCraft_1: x-3673, z2252 = 121
13:35 PirateCraft_1: x-3482, z-115 // PirateCraft_1: x-3472, z-105 = 121
13:35 PirateCraft_1_nether: x3189, z-2747 // PirateCraft_1_nether: x3199, z-2737 = 121
13:35 PirateCraft_1: x-2802, z-1149 // PirateCraft_1: x-2783, z-1131 = 380
13:35 PirateCraft_1: x-4467, z2266 // PirateCraft_1: x-4439, z2303 = 1102
13:35 PirateCraft_1: x-3581, z1610 // PirateCraft_1: x-3571, z1617 = 88
13:35 PirateCraft_1: x-3577, z1619 // PirateCraft_1: x-3564, z1624 = 84
13:35 PirateCraft_1: x-3597, z1556 // PirateCraft_1: x-3592, z1569 = 84
13:35 PirateCraft_1: x-3639, z1535 // PirateCraft_1: x-3633, z1548 = 98
13:35 PirateCraft_1: x-3640, z1575 // PirateCraft_1: x-3633, z1584 = 80
13:35 PirateCraft_1: x-3592, z1599 // PirateCraft_1: x-3573, z1602 = 80
13:35 PirateCraft_1: x7384, z-5512 // PirateCraft_1: x7447, z-5449 = 4096
13:35 PirateCraft_1: x-920, z4527 // PirateCraft_1: x-912, z4535 = 81
13:35 PirateCraft_1: x-3357, z424 // PirateCraft_1: x-3322, z459 = 1296
13:35 PirateCraft_1: x-4060, z2042 // PirateCraft_1: x-4032, z2064 = 667
13:35 PirateCraft_1: x3368, z2271 // PirateCraft_1: x3417, z2285 = 750
13:35 PirateCraft_1_nether: x-2955, z-2439 // PirateCraft_1_nether: x-2945, z-2429 = 121
13:35 PirateCraft_1: x-3580, z2170 // PirateCraft_1: x-3539, z2200 = 1302
13:35 PirateCraft_1_nether: x-3969, z-2999 // PirateCraft_1_nether: x-3959, z-2989 = 121
13:35 PirateCraft_1_nether: x-3178, z-2720 // PirateCraft_1_nether: x-3168, z-2710 = 121
13:35 PirateCraft_1_nether: x-3978, z-2933 // PirateCraft_1_nether: x-3968, z-2923 = 121
13:35 PirateCraft_1: x-3521, z8323 // PirateCraft_1: x-3495, z8369 = 1269
13:35 PirateCraft_1_nether: x-441, z-288 // PirateCraft_1_nether: x-437, z227 = 2580
13:35 PirateCraft_1: x-4981, z3287 // PirateCraft_1: x-4920, z3333 = 2914
13:35 PirateCraft_1_the_end: x-32, z1324 // PirateCraft_1_the_end: x587, z1823 = 310000
13:35 New accrued blocks: 361874
13:35 New total blocks: 533007
13:35 Lawzoneon Accrued blocks: 361874 Bonus blocks: 171133 Group Bonus Blocks: 0
13:35 Total blocks: 533007 Total claimed area: 361874
13:35 Remaining claim blocks to use: 171133 (should be 0)
```

Where it randomly just sets him at _New accrued blocks: 361874_.

So from what I can see, him deleting a players claim only alerted him of the change to his claims blocks.

He notices the claim block change at 2:01 or 3:01 his time.
I noticed this "Adding blocks to fix" at 13:35 my time 14:35 his time.

His username is Lawzoneon, his uuid is 7e556548-0a59-4c43-bd7b-44f3bc9b3261
Full GP Log: https://gist.github.com/FrozenBeard/e1b3a967298c1560a05a3b25e861263c
Full Server Log: https://gist.github.com/FrozenBeard/c7d72de75fadfb798ec14563179ff9fd
His Personal Minecraft Log: https://gist.github.com/FrozenBeard/a624e35d899836ed9493d7cc09404a24

Another note: As I was going through the server logs line by line and checking his GP save progress at
12:41:46 GP removed 109 "accruedblocks" claim blocks from 190850 to 190741 for no reason at all, this accruedblocks should never be negated from, only increment with time played.

**Edit**: **I figured out what GP has done wrong.**
He had 0 claim blocks to use. GP for some reason decided to just remove 109 accrued blocks randomly. accrued blocks should never have anything removed directly from its data. This then triggered the "Adding blocks to fix" check as he had less Claim blocks to what hes actually used due to the -109.
The second problem is the "Adding blocks to fix" isn't right, all it does is just counts up all the Claims hes claimed in the worlds and just resets the "accrued blocks" MySQL data directly to that, completely ignores the bonusblocks. If you look at the previous and current /claimslist on the diffchecker you can actually confirm this, add up all the Claims area and it comes out exactly at 361846, so thats where its getting its number. But if you remove his bonus blocks (171133) from it it gives you 190723, his previous Acquired blocks are 190741, this is very, very close. 

So there are two issues GP is doing here.

- Randomly removing 109 Claim blocks "accruedblocks" from a user.
- The "Fix" check, does not remove the bonusblocks first, it just slaps the FULL claim area into the database

**Comment by RoboMWM** [comment]
> An admin deleted a claim of a user that's been offline for 745 days. The message back to him was him "Claim abandoned. You now have 171108 available claim blocks." 

When you abandon a claim of another player, ~~it will print the available claim blocks for the _claim owner_, not the player running the /abandonclaim command.~~ (Edit: I'm wrong.)

> The second problem is the "Adding blocks to fix" isn't right, all it does is just counts up all the Claims hes claimed in the worlds and just resets the "accrued blocks"

That's how bigscary resolved this issue. There's a config option to disable this fixer though. It's in Advanced->fixNegativeClaimAmounts or something like that (it's the only node under Advanced anyways).

> He had 0 claim blocks to use. GP for some reason decided to just remove 109 accrued blocks randomly. 

Yea, that seems to be what happened. No idea how that occurred at all, are there any addons that touch GP? I'll probably make that value immutable in the API.

I'll look into the files, thanks for those. Issues like these is one of the primary reasons why I want GP v20 to be simple...

**Comment by bigpresh** [comment]
Firstly, a most excellent detailed report, and good troubleshooting figuring out what was going on.

>>  An admin deleted a claim of a user that's been offline for 745 days. The message back to him was him "Claim abandoned. You now have 171108 available claim blocks."

> When you abandon a claim of another player, it will print the available claim blocks for the claim owner, not the player running the /abandonclaim command.

I reckon we should clarify this - so I've done so in PR #323, feedback welcome.

**Comment by TomLewis** [comment]
Well looks like we found 3 bugs with that #323 haha, i'm grateful one has already been solved, thanks @bigpresh .

As for the other two:
The "fix" check, all it needs to do is account for bonusblocks and it'll work, something like:
`allClaimsAddedUp - bonusblocks = fixedTotal, UPDATE bonusblocks = bonusblocks+fixedTotal` the only issue with this is that if the bonusblocks are in minus itll be two minus which goes to a plus, so you would need to check if their bonusblocks are minus first? or maybe theres some better maths, mines terrible!
To be fair this fix shouldn't exist, it should never remove peoples "accrued blocks"so they should never be able to own claims without checking what they have first, and if they don't have enough "accrued blocks" the bloody fix should have added the value onto the god damn BONUS BLOCKS not the "accrued blocks" field. 

As for the vanishing blocks, Directly before he lost the 109 he did a /sellclaimblocks 978, you can see GP do a check on his claim blocks just before he does it, then directly after.
```
[12:41:35] [Thread-364/INFO]: [GriefPrevention] Preparing to save data for 7e556548-0a59-4c43-bd7b-44f3bc9b3261: accrued: 190850 bonus: 172111
[12:41:37] [Server thread/INFO]: Figio22 issued server command: /sellclaim 60
[12:41:37] [Thread-365/INFO]: [GriefPrevention] Preparing to save data for 7eb20905-99ee-4290-ab9d-0eb43e59b05e: accrued: 300 bonus: -219
[12:41:37] [Server thread/INFO]: Lawzoneon issued server command: /back
[12:41:37] [Server thread/INFO]: lazydog11 issued server command: /c its fine do think it whas water and he dos not know all the members and old members
[12:41:37] [Server thread/INFO]: [be] [ Rec  Gov GL] lazydog11 » its fine do think it whas water and he dos not know all the members and old members
[12:41:43] [Server thread/INFO]: Figio22 issued server command: /buyclaim
[12:41:46] [Server thread/INFO]: Lawzoneon issued server command: /sellclaimblocks 978
[12:41:46] [Thread-366/INFO]: [GriefPrevention] Preparing to save data for 7e556548-0a59-4c43-bd7b-44f3bc9b3261: accrued: 190741 bonus: 171133
[12:41:52] [Server thread/INFO]: Lawzoneon issued server command: /bal
```
So maybe its /sellclaimblocks taking that 109 away? not sure why 109 when he used 978, either way it shouldn't take from accrued blocks.

**Comment by TomLewis** [comment]
I wanted to do a quick search to see how many people these affects, I don't keep backups of my GP logs, but I do have 15 log files, so not that many, 15 days of logs from GP.
I scanned them all for "_has more claimed land than blocks available_" to see just how many people roughly these bugs are messing with claim blocks.
**87** matches across 15 files
That's insane, it went from thinking it was a minor bug to a pretty major one.

**Comment by RoboMWM** [comment]
So it turns out my assumption was incorrect, the amount of claim blocks stated in the message is the amount the command executor has.

@FrozenBeard what's your config? What's set for `AbandonReturnRatio`?

**Comment by TomLewis** [comment]
That's a good shout mine is AbandonReturnRatio: 0.9
Sorry for the late reply I'm on holiday!

**Comment by RoboMWM** [comment]
Well that single config option is the cause of all your claimblock problems, lol. Any time someone abandons a claim, their accruedclaimblocks value is adjusted. (No adjustment occurs if `AbandonReturnRatio` is set to 1.0.) Of course, this happens erroneously for administrators who abandon other player's claims (which should be looked into, probably best to put in another issue).

Set it to 1.0 and all these mysterious claim block accrual issues will go away, lol.

**Comment by TomLewis** [comment]
Well im super glad we found the source of the issue, but I was the one that requested that "AbandonReturnRatio" be re-added when GP last got re-written as it plays a large part in survival, can we get it fixed please? I cant just set it to 1.0, not really a fix as it would completely change my servers economy and people will be making claims as temporary places to attack or protect themselves if they are being attacked in PvP to create a quick safe haven, the AbandonReturnRatio deteers from this behavior and is quite important even a 0.1% loss it stops people abusing claims.

**Comment by RoboMWM** [comment]
Yea, the way it looks was as if it was never meant to be there/was leftover from old code. So that explains quite a bit.
